README
RPi-OLED

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc RPi::OLED

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=RPi-OLED

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/RPi-OLED

    CPAN Ratings
        http://cpanratings.perl.org/d/RPi-OLED

    Search CPAN
        http://search.cpan.org/dist/RPi-OLED/


LICENSE AND COPYRIGHT

Copyright (C) 2017 Steve Bertrand

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for RPi-OLED

0.01    Date/time
        First version, released on an unsuspecting world.



lib/RPi/OLED.pm
package RPi::OLED;

use 5.006;
use strict;
use warnings FATAL => 'all';

=head1 NAME

RPi::OLED - The great new RPi::OLED!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use RPi::OLED;

    my $foo = RPi::OLED->new();
    ...

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don't export anything, such as for a purely object-oriented module.

=head1 SUBROUTINES/METHODS

=head2 function1

=cut

sub function1 {
}

=head2 function2

=cut

sub function2 {
}

=head1 AUTHOR

Steve Bertrand, C<< <steveb at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-rpi-oled at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=RPi-OLED>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc RPi::OLED


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=RPi-OLED>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/RPi-OLED>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/RPi-OLED>

=item * Search CPAN

L<http://search.cpan.org/dist/RPi-OLED/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2017 Steve Bertrand.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1; # End of RPi::OLED


lib/RPi/OLED/128x64.pm
package RPi::OLED::128x64;

use warnings;
use strict;

use parent 'RPi::OLED::SSD1306Base';

use Carp qw(croak);
use RPi::OLED::Const;

our $VERSION = '0.01';

sub new {
    my ($class, %args) = @_;

    my $self = $class->SUPER::new(
        width => 128,
        height => 64,
        rst => $args{rst},
        i2c_addr => SSD1306_I2C_ADDRESS
    );

    return $self;
}
sub initialize {
    # 128x64 pixel specific initialization.

    my ($self) = @_;

    $self->command(SSD1306_DISPLAYOFF);
    $self->command(SSD1306_SETDISPLAYCLOCKDIV);
    $self->command(0x80);
    $self->command(SSD1306_SETMULTIPLEX);
    $self->command(0x3F);
    $self->command(SSD1306_SETDISPLAYOFFSET);
    $self->command(0x0);
    $self->command(SSD1306_SETSTARTLINE | 0x0);
    $self->command(SSD1306_CHARGEPUMP);

    if ($self._vccstate == SSD1306_EXTERNALVCC){
        $self->command(0x10);
    }
    else {
        $self->command(0x14);
    }

    $self->command(SSD1306_MEMORYMODE);
    $self->command(0x00);
    $self->command(SSD1306_SEGREMAP | 0x1);
    $self->command(SSD1306_COMSCANDEC);
    $self->command(SSD1306_SETCOMPINS);
    $self->command(0x12);
    $self.command(SSD1306_SETCONTRAST);

    if ($self->{vccstate} == SSD1306_EXTERNALVCC){
        $self->command(0x9F);
    }
    else {
        $self->command(0xCF);
    }

    $self->command(SSD1306_SETPRECHARGE);

    if ($self->{vccstate} == SSD1306_EXTERNALVCC) {
        $self->command(0x22);
    }
    else {
        $self->command(0xF1);
    }

    $self->command(SSD1306_SETVCOMDETECT);
    $self->command(0x40);
    $self->command(SSD1306_DISPLAYALLON_RESUME);
    $self->command(SSD1306_NORMALDISPLAY);
}

1;
__END__

=head1 NAME

RPi::OLED::128x64 - Module for the 128x64 pixel displays

=head1 SYNOPSIS

=head1 DESCRIPTION

=head1 AUTHOR

Steve Bertrand, C<< <steveb at cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2017 Steve Bertrand.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1; # End of RPi::OLED


lib/RPi/OLED/SSD1306Base.pm
package RPi::OLED::SSD1306Base;

use warnings;
use strict;

use Carp qw(croak);
use RPi::Const qw(:all);
use RPi::I2C;
use RPi::OLED::Const;
use RPi::Pin;

our $VERSION = '0.01';

sub new {
    my ($self, %args) = @_;

    # width
    # height
    # rst
    # dc
    # sclk
    # din
    # cs
    # gpio
    # spi
    # i2c
    # i2c_device
    # i2c_addr

    for (keys %args){
        $self->{$_} = $args{$_};
    }

    #FIXME: write sub that handles all i2c args, and dies accordingly if
    # necessary

    if ($self->{i2c}){
        $self->{i2c} = RPi::I2C->new($self->{i2c_addr}, $self->{i2c_device});
    }

    if (defined $self->{rst}){
        $self->{rst} = RPi::Pin->new($self->{rst});
        $self->{rst}->mode(OUTPUT);
    }
    return $self;
}
sub begin {
    # initialize the display

    my ($self, $vcc_state) = @_;
    $self->{vcc_state} = $vcc_state;

    $self->reset;
    $self->initialize;
    $self->command(SSD1306_DISPLAYON);
}
sub command {
    # send command byte to display

    my ($self, $char) = @_;

    if (! defined $char){
        croak "command() requires a byte sent in as an argument\n";
    }
    if ($self->_i2c) {
        my $control = 0x00;
        $self->_i2c($control, $char);
    }
}
sub data {
    # send data byte to display

    my ($self, $char) = @_;

    if (! defined $char){
        croak "data() requires a byte sent in as an argument\n";
    }

    if ($self->_i2c){
        my $control = 0x40;
        $self->_i2c->write_byte($char, $control);
    }
}
sub initialize {
    croak "NOT IMPLEMENTED: You must write an 'initialize()' sub into your " .
          "subclass\n";
}
sub reset {
    # reset the display

    my ($self) = @_;
    return if ! defined $self->{rst};

    $self->{rst}->write(HIGH);
    select(undef, undef, undef, 0.001);
    $self->{rst}->write(LOW);
    select(undef, undef, undef, 0.01);
    $self->{rst}->write(HIGH);
}
sub _i2c {
    return $self->{i2c};
}
1;
__END__

=head1 NAME

RPi::OLED::SSD1306Base - Base class for SSD1306 related OLED displays

=head1 SYNOPSIS

use RPi::OLED::Const

=head1 DESCRIPTION

Provides constant variables for the L<RPi::OLED> distribution.

=head1 AUTHOR

Steve Bertrand, C<< <steveb at cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2017 Steve Bertrand.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1; # End of RPi::OLED


lib/RPi/OLED/Const.pm
package RPi::OLED::Const;

use warnings;
use strict;

our $VERSION = '0.01';

require Exporter;
use base qw( Exporter );
our @EXPORT;

# core constants

use constant {
    SSD1306_I2C_ADDRESS => 0x3C,    # 011110+SA0+RW - 0x3C or 0x3D
    SSD1306_SETCONTRAST => 0x81,
    SSD1306_DISPLAYALLON_RESUME => 0xA4,
    SSD1306_DISPLAYALLON => 0xA5,
    SSD1306_NORMALDISPLAY => 0xA6,
    SSD1306_INVERTDISPLAY => 0xA7,
    SSD1306_DISPLAYOFF => 0xAE,
    SSD1306_DISPLAYON => 0xAF,
    SSD1306_SETDISPLAYOFFSET => 0xD3,
    SSD1306_SETCOMPINS => 0xDA,
    SSD1306_SETVCOMDETECT => 0xDB,
    SSD1306_SETDISPLAYCLOCKDIV => 0xD5,
    SSD1306_SETPRECHARGE => 0xD9,
    SSD1306_SETMULTIPLEX => 0xA8,
    SSD1306_SETLOWCOLUMN => 0x00,
    SSD1306_SETHIGHCOLUMN => 0x10,
    SSD1306_SETSTARTLINE => 0x40,
    SSD1306_MEMORYMODE => 0x20,
    SSD1306_COLUMNADDR => 0x21,
    SSD1306_PAGEADDR => 0x22,
    SSD1306_COMSCANINC => 0xC0,
    SSD1306_COMSCANDEC => 0xC8,
    SSD1306_SEGREMAP => 0xA0,
    SSD1306_CHARGEPUMP => 0x8D,
    SSD1306_EXTERNALVCC => 0x1,
    SSD1306_SWITCHCAPVCC => 0x2,
};

{ # core

    my @const = qw(
        SSD1306_I2C_ADDRESS
        SSD1306_SETCONTRAST
        SSD1306_DISPLAYALLON_RESUME
        SSD1306_DISPLAYALLON
        SSD1306_NORMALDISPLAY
        SSD1306_INVERTDISPLAY
        SSD1306_DISPLAYOFF
        SSD1306_DISPLAYON
        SSD1306_SETDISPLAYOFFSET
        SSD1306_SETCOMPINS
        SSD1306_SETVCOMDETECT
        SSD1306_SETDISPLAYCLOCKDIV
        SSD1306_SETPRECHARGE
        SSD1306_SETMULTIPLEX
        SSD1306_SETLOWCOLUMN
        SSD1306_SETHIGHCOLUMN
        SSD1306_SETSTARTLINE
        SSD1306_MEMORYMODE
        SSD1306_COLUMNADDR
        SSD1306_PAGEADDR
        SSD1306_COMSCANINC
        SSD1306_COMSCANDEC
        SSD1306_SEGREMAP
        SSD1306_CHARGEPUMP
        SSD1306_EXTERNALVCC
        SSD1306_SWITCHCAPVCC
    );

    push @EXPORT, @const;
}

# scrolling constants

use constant {
    SSD1306_ACTIVATE_SCROLL => 0x2F,
    SSD1306_DEACTIVATE_SCROLL => 0x2E,
    SSD1306_SET_VERTICAL_SCROLL_AREA => 0xA3,
    SSD1306_RIGHT_HORIZONTAL_SCROLL => 0x26,
    SSD1306_LEFT_HORIZONTAL_SCROLL => 0x27,
    SSD1306_VERTICAL_AND_RIGHT_HORIZONTAL_SCROLL => 0x29,
    SSD1306_VERTICAL_AND_LEFT_HORIZONTAL_SCROLL => 0x2A,
};

{ # scrolling

    my @const = qw(
        SSD1306_ACTIVATE_SCROLL
        SSD1306_DEACTIVATE_SCROLL
        SSD1306_SET_VERTICAL_SCROLL_AREA
        SSD1306_RIGHT_HORIZONTAL_SCROLL
        SSD1306_LEFT_HORIZONTAL_SCROLL
        SSD1306_VERTICAL_AND_RIGHT_HORIZONTAL_SCROLL
        SSD1306_VERTICAL_AND_LEFT_HORIZONTAL_SCROLL
    );

    push @EXPORT, @const;
}

__END__

=head1 NAME

RPi::OLED::Const - Helper module for RPi::OLED to provide constant variables

=head1 SYNOPSIS

use RPi::OLED::Const

=head1 DESCRIPTION

Provides constant variables for the L<RPi::OLED> distribution.

=head1 EXPORT

This module exports all related constant variables present in this module. It
does not export any subs.

=head1 AUTHOR

Steve Bertrand, C<< <steveb at cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2017 Steve Bertrand.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1; # End of RPi::OLED
